### PR TITLE
Introduce statefulset for OTEL collector

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

This PR implmenets new Reconciler: `otel.reconciler` to automate a StatefulSet for the OpenTelemetry collector for extensions.

[K8S-10311](https://dt-rnd.atlassian.net/browse/K8S-10311)

> NOTE:
>  dynatrace-extensions-collector-0     ●  0/1   CrashLoopBackOff
It's okey for now, we wait some fixes from EEC team to release new image which supports fetching config via endpoint.

## How can this be tested?

Deploy dynakube with prometheus.enabled = true

- Make sure `dynatrace-extensions-collector` statefulset is presented and has all labels/annotations.
- Also, env variables are presented for the Otelc container.

Unfortunately, you can't test more.

### CODE REVIEW

**Review by aorcholski — COMMENTED** [review]
`pod.labels.pod-index` label doesn't exist. I checked EEC,OTELC,AG pods

**Comment by andriisoldatenko** [comment]
> please add:
> 
> ```
>   serviceAccount: dynatrace-extensions-collector
>   serviceAccountName: dynatrace-extensions-collector
> ```

done

**Review by wepudt — COMMENTED** [review]
I tried putting on labels as stated in the Jira Issue using a template, but failed. I also cannot find an entry in the CRD (analog to `extensionExecutionController`)- am I missing something?

**Comment by andriisoldatenko** [comment]
> I tried putting on labels as stated in the Jira Issue using a template, but failed. I also cannot find an entry in the CRD (analog to `extensionExecutionController`)- am I missing something?

replied in DM, but it works.

```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dynakube
  namespace: dynatrace
  annotations:
    feature.dynatrace.com/ignored-namespaces: '["^kube-.*", "openshift(-.*)?", "^gke-.*", "^gmp-.*" ]'
    dynatrace.com/extensions: '{"prometheus":{"enabled":true}}'
spec:
  apiUrl: https://hbc47170.dev.dynatracelabs.com/api
  customPullSecret: azurecr
  metadataEnrichment:
    enabled: true
  extensions:
    prometheus:
      enabled: true
  templates:
    extensionExecutionController:
      imageRef:
        repository: extk8sregistry.azurecr.io/eec/dynatrace-eec
        tag: 1.289.0.20240311-112000
    openTelemetryCollector:
      labels:
        test-label: test-value
```

```
➜  dynatrace-operator git:(feature/otel-statefulset) k describe pod dynatrace-extensions-collector-0 | rg -C10 Labels
Name:             dynatrace-extensions-collector-0
Namespace:        dynatrace
Priority:         0
Service Account:  dynatrace-extensions-collector
Node:             gke-andrii-test-default-pool-cf2748c9-0tv0/10.128.0.61
Start Time:       Wed, 07 Aug 2024 11:04:23 +0200
Labels:           app.kubernetes.io/component=dynatrace-extensions-collector
                  app.kubernetes.io/created-by=dynakube
                  app.kubernetes.io/managed-by=dynatrace-operator
                  app.kubernetes.io/name=dynatrace-extensions-collector
                  app.kubernetes.io/version=0.0.0
                  apps.kubernetes.io/pod-index=0
                  controller-revision-hash=dynatrace-extensions-collector-76475b8f6c
                  statefulset.kubernetes.io/pod-name=dynatrace-extensions-collector-0
                  test-label=test-value <---- MY LABEL
```

File: pkg/controllers/dynakube/extension/eec/statefulset.go [hunk 1]
```diff
@@ -57,7 +57,7 @@ func (r *reconciler) createOrUpdateStatefulset(ctx context.Context) error {
 		statefulset.SetAllAnnotations(nil, r.dk.Spec.Templates.ExtensionExecutionController.Annotations),
 		statefulset.SetAffinity(buildAffinity()),
 		statefulset.SetTolerations(r.dk.Spec.Templates.ExtensionExecutionController.Tolerations),
-		statefulset.SetTopologySpreadConstraints(buildTopologySpreadConstraints(r.dk.Spec.Templates.ExtensionExecutionController.TopologySpreadConstraints, r.dk.Name)),
+		statefulset.SetTopologySpreadConstraints(utils.BuildTopologySpreadConstraints(r.dk.Spec.Templates.ExtensionExecutionController.TopologySpreadConstraints, appLabels)),
 		statefulset.SetServiceAccount(serviceAccountName),
 		statefulset.SetSecurityContext(buildPodSecurityContext()),
 		statefulset.SetUpdateStrategy(buildUpdateStrategy()),
```
Comment (aorcholski) [inline pkg/controllers/dynakube/extension/eec/statefulset.go:63]: Could you replace EEC `buildUpdateStrategy` with `utils.BuildUpdateStrategy` and remove `buildUpdateStrategy` ?

File: pkg/controllers/dynakube/extension/otel/statefulset_test.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+				Effect:   corev1.TaintEffectNoSchedule,
+			},
+		}
+		dk.Spec.Templates.OpenTelemetryCollector.Tolerations = customTolerations
+
+		statefulSet := getStatefulset(t, dk)
+
+		assert.Equal(t, customTolerations, statefulSet.Spec.Template.Spec.Tolerations)
+	})
+}
```
Comment (aorcholski) [inline pkg/controllers/dynakube/extension/otel/statefulset_test.go:308]: Please add TestSecurityContext and TestUpdateStrategy functions to make sure we do set them.

File: pkg/util/kubeobjects/statefulset/builder.go [hunk 3]
```diff
@@ -31,7 +31,10 @@ func Build(owner metav1.Object, name string, container corev1.Container, options
 
 func SetReplicas(replicas int32) builder.Option[*appsv1.StatefulSet] {
 	return func(s *appsv1.StatefulSet) {
-		s.Spec.Replicas = address.Of(replicas)
+		s.Spec.Replicas = address.Of(int32(1))
+		if replicas != 0 {
```
Comment (aorcholski) [inline pkg/util/kubeobjects/statefulset/builder.go:35]: IMHO it should be possible to set Replicas to 0 , it's hard to do that by mistake

Comment (andriisoldatenko) [reply pkg/util/kubeobjects/statefulset/builder.go:35]: okey.

File: pkg/controllers/dynakube/extension/eec/statefulset.go (hunk 1 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/extension/eec/statefulset.go]: fixed in 5f946433

File: pkg/controllers/dynakube/extension/otel/statefulset_test.go (hunk 2 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/extension/otel/statefulset_test.go:308]: fixed in 1767f67d

**Review by wepudt — APPROVED** [review]
lgtm

code look nice, tested on my cluster with labels and annotations defined in template, checked for env vars.

File: pkg/util/kubeobjects/statefulset/builder.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/util/kubeobjects/statefulset/builder.go]: fixed in 19c750df

**Review by aorcholski — APPROVED** [review]
LGTM

**Review by wepudt — APPROVED** [review]
lgtm

**Pull request merged**